README
RPC::ExtDirect::Server
======================

    This module implements a minimal Ext.Direct capable server in pure Perl.
    Its main purpose is to be used as a lightweight drop-in replacement for
    more complex production environments like Plack or Apache/mod_perl, e.g.
    for testing and mockups.

    It can also be used successfully as the basis for production application
    servers when feature richness is not a requirement, or when backwards
    compatibility is the primary concern. Another possible application for
    RPC::ExtDirect::Server is embedded systems; due to its small footprint
    and pure Perl dependencies it can be used easily in resource constrained
    platforms.

    Finally, many legacy computing systems can be retrofitted with modern
    HTML5 Web interface applications with Ext.Direct back end built on top
    of RPC::ExtDirect::Server; it is compatible with (and is tested
    against!) Perl 5.6 and newer.

INSTALLATION

    To install this module type the following:

        perl Makefile.PL
        make && make test
        make install

COPYRIGHT AND LICENSE

    Copyright (c) 2012-2015 Alex Tokarev <tokarev@cpan.org>.

    Portions of the code that were taken from HTTP::Date are copyright (c)
    1995-1999, Gisle Aas.

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. See perlartistic.



Changes
1.24  Wed Oct 12 01:07:16 PDT 2016
    - No longer use CGI::Simple if CGI.pm > 4.0 is installed
    - Refactored Server.pm not to use `each` hash iterator
    - Updated core dependencies to version 3.24

1.21  Mon Oct 10 19:42:43 PDT 2016
    - Do not apply older CGI.pm compatibility patch with HTTP::Server::Simple >= 0.51
    - Do not use File::MimeInfo if shared-mime-info database is missing
    - Misc documentation fixes

1.20  Thu May 14 20:50:02 PDT 2015
    - Updated to support metadata handling feature in RPC::ExtDirect 3.20+

1.01  Fri Jun 20 20:51:34 2014
    - Better compatibility with CGI.pm < 3.36
    - Test scripts will clean %ENV to avoid connecting to proxies
    - Minor fixes

1.00  Thu Jun 12 21:32:00 2014
    - Updated to accommodate for the changes in RPC::ExtDirect 3.0 and
      CGI::ExtDirect 3.0
    - maybe_start_server for @ARGV handling, and other helpful stuff
    - Many bug fixes and improvements
    - Tons of documentation added
    - More tests that use HTTP::Tiny now instead of WWW::Mechanize
    - Most of the dependencies made optional

0.01  Wed Jun 20 10:42:25 2012
	- original version; code works and tests pass but lacks documentation



t/server.pl
# This script can be used for manual HTTP server testing in case
# something goes awry

use strict;
use warnings;

use RPC::ExtDirect::Server;

my $port = shift @ARGV || 30000 + int rand 10000;

my $server = RPC::ExtDirect::Server->new(
    static_dir => '/tmp',
    port       => $port,
);
$port = $server->port;

print "Listening on port $port\n";

$server->run();



t/lib/RPC/ExtDirect/Server/Test/Util.pm
package RPC::ExtDirect::Server::Test::Util;

use Exporter;
use Test::More;
use HTTP::Tiny;

use base 'Exporter';

our @EXPORT = qw/
    get
    post
    is_status
    is_content
    like_content
    is_header
    like_header
/;

sub get {
    my ($uri, $opt, $arg) = @_;

    my $http = HTTP::Tiny->new( max_redirect => 0, %$arg, );

    # HTTP::Tiny is picky about its arguments
    return $http->get($uri, $opt ? ($opt) : ());
}

sub post {
    my ($uri, $opt, $arg) = @_;

    my $http = HTTP::Tiny->new( max_redirect => 0, %$arg, );

    return $http->post($uri, $opt);
}

sub is_status {
    my ($r, $want, $msg) = @_;

    is $r->{status}, $want, $msg or diag explain "Response:", $resp;
}

sub is_content {
    my ($r, $want, $msg) = @_; 

    is $r->{content}, $want, $msg or diag explain "Response:", $resp;
}

sub like_content {
    my ($r, $want, $msg) = @_;

    like $r->{content}, $want, $msg or diag explain "Response:", $resp;
}

sub is_header {
    my $r   = shift;
    my $hdr = lc shift;
    my ($want, $msg) = @_;

    is $r->{headers}->{$hdr}, $want, $msg
        or diag explain "Response:", $resp;
}

sub like_header {
    my $r   = shift;
    my $hdr = lc shift;
    my ($want, $msg) = @_;

    like $r->{headers}->{$hdr}, $want, $msg
        or diag explain "Response:", $resp;
}

1;


lib/RPC/ExtDirect/Server.pm
package RPC::ExtDirect::Server;

use strict;
use warnings;
no  warnings 'uninitialized';   ## no critic

use Carp;

use RPC::ExtDirect::Util::Accessor;
use RPC::ExtDirect::Config;
use RPC::ExtDirect::API;
use RPC::ExtDirect;
use CGI::ExtDirect;

use HTTP::Server::Simple::CGI;
use base 'HTTP::Server::Simple::CGI';

### PACKAGE GLOBAL VARIABLE ###
#
# Version of this module.
#

our $VERSION = '1.21';

# We're trying hard not to depend on any non-core modules,
# but there's no reason not to use them if they're available
my ($have_http_date, $have_cgi_simple);

{
    local $@;
    $have_http_date  = eval "require HTTP::Date";

    # CGI::Simple is only meaningful if we're using old CGI.pm,
    # and only if certain version of CGI::Simple is available
    $have_cgi_simple = $CGI::VERSION < 4.0
        && eval "require CGI::Simple; $CGI::Simple::VERSION > 1.113";
}

# CGI.pm < 3.36 does not support HTTP_COOKIE environment variable
# with multiple values separated by commas instead of semicolons,
# which is exactly what HTTP::Server::Simple::CGI::Environment
# does in version <= 0.51. The module below will fix that.

if ( $CGI::VERSION < 3.36 && $HTTP::Server::Simple::VERSION <= 0.51 ) {
    local $@;

    require RPC::ExtDirect::Server::Patch::HTTPServerSimple;
}

# We assume that HTTP::Date::time2str is better maintained,
# so use it if we can. If HTTP::Date is not installed,
# fall back to our own time2str - which was shamelessly copied
# from HTTP::Date anyway.
if ( $have_http_date ) {
    *time2str = *HTTP::Date::time2str;
    *str2time = *HTTP::Date::str2time;
}
else {
    eval <<'END_SUB';
    my @DoW = qw(Sun Mon Tue Wed Thu Fri Sat);
    my @MoY = qw(Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec);
    
    sub time2str {
        my $time = shift;
        
        $time = time unless defined $time;
        
        my ($sec, $min, $hour, $mday, $mon, $year, $wday)
            = gmtime($time);
        
        return sprintf "%s, %02d %s %04d %02d:%02d:%02d GMT",
                       $DoW[$wday],
                       $mday,
                       $MoY[$mon],
                       $year + 1900,
                       $hour,
                       $min,
                       $sec
                       ;
    }
END_SUB
}

my %DEFAULTS = (
    index_file    => 'index.html',
    expires_after => 259200, # 3 days in seconds
    buffer_size   => 262144, # 256kb in bytes
);

### PUBLIC CLASS METHOD (CONSTRUCTOR) ###
#
# Instantiate a new HTTPServer
#

sub new {
    my ($class, %arg) = @_;

    my $api        = delete $arg{api}        || RPC::ExtDirect->get_api();
    my $config     = delete $arg{config}     || $api->config;
    my $host       = delete $arg{host}       || '127.0.0.1';
    my $port       = delete $arg{port}       || 8080;
    my $cust_disp  = delete $arg{dispatch}   || [];
    my $static_dir = delete $arg{static_dir} || '/tmp';
    my $cgi_class  = delete $arg{cgi_class};

    $config->set_options(%arg);

    my $self = $class->SUPER::new($port);
    
    $self->_init_cgi_class($cgi_class);
    
    $self->api($api);
    $self->config($config);
    $self->host($host);

    $self->static_dir($static_dir);
    $self->logit("Using static directory ". $self->static_dir);
    
    foreach my $k (keys %DEFAULTS) {
        my $v = $DEFAULTS{$k};
        my $value = exists $arg{ $k } ? delete $arg{ $k } : $v;
        
        $self->$k($value);
    }

    $self->_init_dispatch($cust_disp);
    
    return bless $self, $class;
}

### PUBLIC INSTANCE METHOD ###
#
# Find matching method by URI and dispatch it.
# This is an entry point for HTTP::Server::Simple API, and is called
# by the underlying module (in fact HTTP::Server::Simple::CGI).
#

sub handle_request {
    my ($self, $cgi) = @_;
    
    my $path_info = $cgi->path_info();
    
    my $debug = $self->config->debug;
    
    $self->logit("Handling request: $path_info") if $debug;
    
    $cgi->nph(1);
    
    HANDLER:
    for my $handler ( @{ $self->dispatch } ) {
        my $match = $handler->{match};
        
        $self->logit("Matching '$path_info' against $match") if $debug;
        
        next HANDLER unless $path_info =~ $match;
        
        $self->logit("Got specific handler with match '$match'") if $debug;
        
        my $code = $handler->{code};
        
        # Handlers are always called as if they were ref($self)
        # instance methods
        return $code->($self, $cgi);
    }

    $self->logit("No specific handlers found, serving default") if $debug;
    
    return $self->handle_default($cgi, $path_info);
}

### PUBLIC INSTANCE METHOD ###
#
# Default request handler
#

sub handle_default {
    my ($self, $cgi, $path) = @_;

    # Lame security measure
    return $self->handle_403($cgi, $path) if $path =~ m{/\.\.};

    my $static = $self->static_dir();
    $static   .= '/' unless $path =~ m{^/};

    my $file_name = $static . $path;
    
    my $file_exists   = -f $file_name;
    my $file_readable = -r $file_name;

    if ( -d $file_name ) {
        $self->logit("Got directory request");
        return $self->handle_directory($cgi, $path);
    }
    elsif ( $file_exists && !$file_readable ) {
        $self->logit("File exists but no permissions to read it (403)");
        return $self->handle_403($cgi, $path);
    }
    elsif ( $file_exists && $file_readable ) {
        $self->logit("Got readable file, serving as static content");
        return $self->handle_static(
            cgi       => $cgi,
            path      => $path,
            file_name => $file_name,
        );
    }
    else {
        return $self->handle_404($cgi, $path);
    };

    return 1;
}

### PUBLIC INSTANCE METHOD ###
#
# Handle directory request. Usually results in a redirect
# but can be overridden to do something fancier.
#

sub handle_directory {
    my ($self, $cgi, $path) = @_;
    
    # Directory requested, redirecting to index.html
    $path =~ s{/+$}{};
    
    my $index_file = $self->index_file;
    
    $self->logit("Redirecting to $path/$index_file");
    
    my $out = $self->stdio_handle;

    print $out $cgi->redirect(
        -uri    => "$path/$index_file",
        -status => '301 Moved Permanently'
    );
    
    return 1;
}

### PUBLIC INSTANCE METHOD ###
#
# Handle static content
#

sub handle_static {
    my ($self, %arg) = @_;

    my $cgi       = $arg{cgi};
    my $file_name = $arg{file_name};

    $self->logit("Handling static request for $file_name");

    my ($fsize, $fmtime) = (stat $file_name)[7, 9];
    my ($type, $charset) = $self->_guess_mime_type($file_name);
    
    $self->logit("Got MIME type $type");
    
    my $out = $self->stdio_handle;
    
    # We're only processing If-Modified-Since if HTTP::Date is installed.
    # That's because str2time is not trivial and there's no point in
    # copying that much code. The feature is not worth it.
    if ( $have_http_date ) {
        my $ims = $cgi->http('If-Modified-Since');
    
        if ( $ims && $fmtime <= str2time($ims) ) {
            $self->logit("File has not changed, serving 304");
            print $out $cgi->header(
                -type   => $type,
                -status => '304 Not Modified',
            );
        
            return 1;
        };
    }
    
    my ($in, $buf);

    if ( not open $in, '<', $file_name ) {
        $self->logit("File is unreadable, serving 403");
        return $self->handle_403($cgi);
    };

    $self->logit("Serving file content with 200");
    
    my $expires = $self->expires_after;

    print $out $cgi->header(
        -type    => $type,
        -status  => '200 OK',
        -charset => ($charset || ($type !~ /image|octet/ ? 'utf-8' : '')),
        ( $expires ? ( -Expires => time2str(time + $expires) ) : () ),
        -Content_Length => $fsize,
        -Last_Modified  => time2str($fmtime),
    );

    my $bufsize = $self->buffer_size;
    
    binmode $in;
    binmode $out;
    
    # Making the out handle hot helps in older Perls
    {
        my $orig_fh = select $out;
        $| = 1;
        select $orig_fh;
    }

    print $out $buf while sysread $in, $buf, $bufsize;

    return 1;
}

### PUBLIC INSTANCE METHOD ###
#
# Return Ext.Direct API declaration JavaScript
#

sub handle_extdirect_api {
    my ($self, $cgi) = @_;

    $self->logit("Got Ext.Direct API request");

    return $self->_handle_extdirect($cgi, 'api');
}

### PUBLIC INSTANCE METHOD ###
#
# Route Ext.Direct method calls
#

sub handle_extdirect_router {
    my ($self, $cgi) = @_;

    $self->logit("Got Ext.Direct route request");

    return $self->_handle_extdirect($cgi, 'route');
}

### PUBLIC INSTANCE METHOD ###
#
# Poll Ext.Direct event providers for events
#

sub handle_extdirect_poll {
    my ($self, $cgi) = @_;

    $self->logit("Got Ext.Direct event poll request");

    return $self->_handle_extdirect($cgi, 'poll');
}

### PUBLIC INSTANCE METHOD ###
#
# Return 403 header without a body.
#

sub handle_403 {
    my ($self, $cgi, $uri) = @_;
    
    $self->logit("Handling 403 for URI $uri");
    
    my $out = $self->stdio_handle;
    
    print $out $cgi->header(-status => '403 Forbidden');
    
    return 1;
}

### PUBLIC INSTANCE METHOD ###
#
# Return 404 header without a body.
#

sub handle_404 {
    my ($self, $cgi, $uri) = @_;

    $self->logit("Handling 404 for URI $uri");
    
    my $out = $self->stdio_handle;

    print $out $cgi->header(-status => '404 Not Found');

    return 1;
}

### PUBLIC INSTANCE METHOD ###
#
# Log debugging info to STDERR
#

sub logit {
    my $self = shift;
    
    print STDERR @_, "\n" if $self->config->debug;
}

### PUBLIC PACKAGE SUBROUTINE ###
#
# Prints banner, but only if debugging is on
#

sub print_banner {
    my $self = shift;

    $self->SUPER::print_banner if $self->config->debug;
}

### PUBLIC INSTANCE METHODS ###
#
# Read-write accessors
#

RPC::ExtDirect::Util::Accessor->mk_accessors(
    simple => [qw/
        api
        config
        dispatch
        static_dir
        index_file
        expires_after
        buffer_size
    /],
);

############## PRIVATE METHODS BELOW ##############

### PRIVATE INSTANCE METHOD ###
#
# Parse HTTP request line. Returns three values: request method,
# URI and protocol.
#
# This method is overridden to improve parsing speed. The original
# method is reading characters from STDIN one by one, which
# results in abysmal performance. Not sure what was the intent
# there but I haven't encountered any problems so far with the
# faster implementation below.
#
# The same is applicable to the parse_headers() below.
#

sub parse_request {
    my $self = shift;

    my $io_handle = $self->stdio_handle;
    my $input     = <$io_handle>;

    return unless $input;

    $input =~ /^(\w+)\s+(\S+)(?:\s+(\S+))?\r?$/ and
        return ( $1.'', $2.'', $3.'' );
}

### PRIVATE INSTANCE METHOD ###
#
# Parse incoming HTTP headers from input file handle and return
# an arrayref of header/value pairs.
#

sub parse_headers {
    my $self = shift;

    my $io_handle = $self->stdio_handle;

    my @headers;

    while ( my $input = <$io_handle> ) {
        $input =~ s/[\r\l\n\s]+$//;
        last if !$input;

        push @headers, $1 => $2
            if $input =~ /^([^()<>\@,;:\\"\/\[\]?={} \t]+):\s*(.*)/i;
    };

    return \@headers;
}

### PRIVATE INSTANCE METHOD ###
#
# Initialize CGI class. Used by constructor.
#

sub _init_cgi_class {
    my ($self, $cgi_class) = @_;
    
    # Default to CGI::Simple > 1.113 if it's available, unless the user
    # overrode cgi_class to do something else. CGI::Simple 1.113 and
    # earlier has a bug with form/multipart file upload handling, so
    # we don't use it even if it is installed.
    if ( $cgi_class ) {
        $self->cgi_class($cgi_class);
        
        if ( $cgi_class eq 'CGI' ) {
            $self->cgi_init(sub {
                local $@;
                
                eval {
                    require CGI;
                    CGI::initialize_globals();
                }
            });
        }
        else {
            $self->cgi_init(sub {
                eval "require $cgi_class";
            });
        }
    }
    elsif ( $have_cgi_simple && $self->cgi_class eq 'CGI' ) {
        $self->cgi_class('CGI::Simple');
        $self->cgi_init(undef);
    }
}

### PRIVATE INSTANCE METHOD ###
#
# Initialize dispatch table. Used by constructor.
#

sub _init_dispatch {
    my ($self, $cust_disp) = @_;
    
    my $config = $self->config;
    
    my @dispatch;

    # Set the custom handlers so they would come first served.
    # Format:
    # [ qr{URI} => \&method, ... ]
    # [ { match => qr{URI}, code => \&method, } ]
    while ( my $uri = shift @$cust_disp ) {
        $self->logit("Installing custom handler for URI: $uri");
        push @dispatch, {
            match => qr{$uri},
            code  => shift @$cust_disp,
        };
    };
    
    # The default Ext.Direct handlers always come last
    for my $type ( qw/ api router poll / ) {
        my $uri_getter = "${type}_path";
        my $handler    = "handle_extdirect_${type}";
        my $uri        = $config->$uri_getter;
        
        if ( $uri ) {
            push @dispatch, {
                match => qr/^\Q$uri\E$/, code => \&{ $handler },
            }
        }
    }

    $self->dispatch(\@dispatch);
}

### PRIVATE INSTANCE METHOD ###
#
# Do the actual heavy lifting for Ext.Direct calls
#

sub _handle_extdirect {
    my ($self, $cgi, $what) = @_;

    my $exd = CGI::ExtDirect->new({
        api    => $self->api,
        config => $self->config,
        cgi    => $cgi,
    });

    # Standard CGI headers for this handler
    my %std_cgi = ( '-nph' => 1, '-charset' => 'utf-8' );
    
    my $out = $self->stdio_handle;

    print $out $exd->$what( %std_cgi );

    return 1;
}

# Popular MIME types, taken from http://lwp.interglacial.com/appc_01.htm
my %MIME_TYPES = (
    au   => 'audio/basic',
    avi  => 'vide/avi',
    bmp  => 'image/bmp',
    bz2  => 'application/x-bzip2',
    css  => 'text/css',
    dtd  => 'application/xml-dtd',
    doc  => 'application/msword',
    gif  => 'image/gif',
    gz   => 'application/x-gzip',
    ico  => 'image/x-icon',
    hqx  => 'application/mac-binhex40',
    htm  => 'text/html',
    html => 'text/html',
    jar  => 'application/java-archive',
    jpg  => 'image/jpeg',
    jpeg => 'image/jpeg',
    js   => 'text/javascript',
    json => 'application/json',
    midi => 'audio/x-midi',
    mp3  => 'audio/mpeg',
    mpeg => 'video/mpeg',
    ogg  => 'audio/vorbis',
    pdf  => 'application/pdf',
    pl   => 'application/x-perl',
    png  => 'image/png',
    ppt  => 'application/vnd.ms-powerpoint',
    ps   => 'application/postscript',
    qt   => 'video/quicktime',
    rdf  => 'application/rdf',
    rtf  => 'application/rtf',
    sgml => 'text/sgml',
    sit  => 'application/x-stuffit',
    svg  => 'image/svg+xml',
    swf  => 'application/x-shockwave-flash',
    tgz  => 'application/x-tar',
    tiff => 'image/tiff',
    tsv  => 'text/tab-separated-values',
    txt  => 'text/plain',
    wav  => 'audio/wav',
    xls  => 'application/excel',
    xml  => 'application/xml',
    zip  => 'application/zip',
);

### PRIVATE INSTANCE METHOD ###
#
# Return the guessed MIME type for a file name
#

# We try to use File::LibMagic or File::MimeInfo if available
{
    local $@;
    
    my $have_libmagic = $ENV{DEBUG_NO_FILE_LIBMAGIC}
                      ? !1
                      : eval "require File::LibMagic";
    
    #
    # File::MimeInfo is a bit kludgy: it depends on shared-mime-info database
    # being installed, and when said database is missing it will do only
    # very basic guessing that is not very useful. Not only that, it will
    # also complain loudly into STDERR about the missing database, which is
    # definitely not helping either. So in addition to checking if the module
    # itself is available we poke a bit deeper and decide if it's worth using.
    #
    my $have_mimeinfo = !$ENV{DEBUG_NO_FILE_MIMEINFO} &&
        eval {
            require File::MimeInfo;
            # This is a dependency of File::MimeInfo
            require File::BaseDir;

            # When both arrays are empty the module is essentially useless
            @File::MimeInfo::DIRS || File::BaseDir::data_files('mime/globs');
        };
    
    sub _guess_mime_type {
        my ($self, $file_name) = @_;
        
        my ($type, $charset);
        
        if ( $have_libmagic ) {
            my $magic = File::LibMagic->new();
            my $mime = $magic->checktype_filename($file_name);
            
            ($type, $charset) = $mime =~ m{^([^;]+);\s*charset=(.*)$};
        }
        elsif ( $have_mimeinfo ) {
            my $mimeinfo = File::MimeInfo->new();
            $type = $mimeinfo->mimetype($file_name);
        }
        
        # If none of the advanced modules are present, resort to
        # guesstimating by file extension
        else {
            my ($suffix) = $file_name =~ /.*\.(\w+)$/;
            
            $type = $MIME_TYPES{ $suffix };
        }
        
        return ($type || 'application/octet-stream', $charset);
    }
}

1;


lib/RPC/ExtDirect/Server/Util.pm
package RPC::ExtDirect::Server::Util;

use strict;
use warnings;
no  warnings 'uninitialized';   ## no critic

use Carp;
use Socket;
use Getopt::Std;
use Exporter;

use RPC::ExtDirect::Server;

use base 'Exporter';

our @EXPORT = qw/
    maybe_start_server
    start_server
    stop_server
/;

### PRIVATE PACKAGE SUBROUTINES ###
#
# Internal use only.
#

{
    my ($server_pid, $server_host, $server_port);
    
    sub get_server_pid { $server_pid };
    sub set_server_pid { $server_pid = shift; };
    
    sub get_server_host { $server_host };
    sub set_server_host { $server_host = shift };
    
    sub get_server_port { $server_port };
    sub set_server_port { $server_port = shift; };
}

### EXPORTED PUBLIC PACKAGE SUBROUTINE ###
#
# See if a host and port were given in the @ARGV, and start a new
# server instance if not.
#

sub maybe_start_server {
    if ( @ARGV ) {
        my %opt;
        
        getopts('h:p:fes:t:', \%opt);
        
        # If a port is given but not the host name, we assume localhost
        $opt{h} ||= '127.0.0.1';
        
        return ($opt{h}, $opt{p}) if $opt{p};
        
        # Not quoting $opt{s} makes my text editor lose its mind ;)
        push @_, static_dir => $opt{'s'} if $opt{'s'};
        push @_, foreground => 1         if $opt{f};
        push @_, enbugger   => 1         if $opt{e};
        push @_, set_timer  => $opt{t}   if defined $opt{t};    
    }
    
    return start_server( @_ );
}

### EXPORTED PUBLIC PACKAGE SUBROUTINE ###
#
# Start an RPC::ExtDirect::Server instance, wait for it to bind
# to a port and return the host and port number.
# If an instance has already been started, return its parameters
# instead of starting a new one.
#

sub start_server {
    my (%arg) = @_;
    
    {
        my $host = get_server_host;
        my $port = get_server_port;
        
        if ( $port ) {
            return wantarray ? ($host, $port)
                 :             "$host:$port"
                 ;
        }
    }
    
    # This parameter is used for internal testing
    my $sleep      = delete $arg{sleep};
    my $foreground = delete $arg{foreground};
    my $enbugger   = delete $arg{enbugger};
    my $set_timer  = delete $arg{set_timer};
    my $timeout    = delete $arg{timeout} || 30;
    
    # We default to verbose exceptions, which is against Ext.Direct spec
    # but feels somewhat saner and is better for testing
    $arg{verbose_exceptions} = 1 unless defined $arg{verbose_exceptions};
    
    if ( $enbugger ) {
        local $@;
        eval "require Enbugger";
    }
    
    # Interactive start means we're not forking but running the server
    # in the current process. Useful for Enbugging.
    if ( $foreground ) {
        if ( $set_timer ) {
            $SIG{ALRM} = sub { Enbugger->stop; };
            
            alarm $set_timer;
        }
        
        do_start_server(
            %arg,
            
            after_listener => sub {
                my ($self) = @_;
                
                my $host = $self->host;
                my $port = $self->port;
                
                print ref($self)." is listening on $host:$port\n";
            }
        );
        
        # This should be unreachable, but just in case
        exit 0;
    }

    my ($pid, $pipe_rd, $pipe_wr);
    pipe($pipe_rd, $pipe_wr) or die "Can't open pipe: $!";

    if ( $pid = fork ) {
        close $pipe_wr;
        local $SIG{CHLD} = sub { waitpid $pid, 0 };

        # Wait until the kid starts up, but don't block forever either
        my ($host, $port) = eval {
            local $SIG{ALRM} = sub { die "alarm\n" };
            alarm $timeout;
            
            my ($host, $port) = split /:/, <$pipe_rd>;
            close $pipe_rd;
            
            alarm 0;
            
            ($host, $port + 0); # Easier than chomp
        };
        
        if ( my $err = $@ ) {
            # If timed out, try to clean up the kid anyway
            eval { kill 2, $pid };
            
            croak $err eq "alarm\n" ? "Timed out waiting for " .
                                      "the server instance to start " .
                                      "after $timeout seconds"
                :                     $err
                ;
        }
        
        set_server_pid($pid);
        set_server_host($host);
        set_server_port($port);

        return wantarray ? ($host, $port)
             :             "$host:$port"
             ;
    }
    elsif ( defined $pid && $pid == 0 ) {
        close $pipe_rd;

        srand;
        
        sleep $sleep if $sleep;
        
        do_start_server(
            %arg,
            
            # TODO This is a dirty hack - find a better way of
            # injecting after_setup_listener. Maybe send a patch
            # to HTTP::Server::Simple maintainer to make this easier?
            after_listener => sub {
                my $self = shift;
        
                my $host = inet_ntoa inet_aton $self->host;
                my $port = $self->port;

                print $pipe_wr "$host:$port\n";
                close $pipe_wr;
                
                my $after_setup_listener
                    = $self->{_old_after_setup_listener};
        
                $after_setup_listener->($self, @_)
                    if $after_setup_listener;
            }
        );

        # Should be unreachable, just in case
        exit 0;
    }
    else {
        croak "Can't fork: $!";
    };

    return;
}

### EXPORTED PUBLIC PACKAGE SUBROUTINE ###
#
# Stop previously started server instance
#

sub stop_server {
    my ($pid) = @_;

    $pid = get_server_pid unless defined $pid;

    kill 2, $pid if defined $pid;

    set_server_port(undef);
    set_server_pid(undef);
}

### PRIVATE PACKAGE SUBROUTINE ###
#
# Try to start the server, re-rolling port randomizer
# if the old port is taken
#

sub do_start_server {
    my (%arg) = @_;
    
    my $forced_port    = defined $arg{port};
    my $after_listener = delete $arg{after_listener};
    my $server_class   = delete $arg{server_class} ||
                                'RPC::ExtDirect::Server';

    if ( !$forced_port ) {
       $arg{port} = random_port();
    }

    my $server = $server_class->new(%arg);
    
    if ( $after_listener ) {
        $server->{_old_after_setup_listener}
            = $server_class->can('after_setup_listener');

        no strict 'refs';
        *{$server_class.'::after_setup_listener'} = $after_listener;
    }

    # If the port is taken, reroll the random generator and try again
    do {
        eval { $server->run() };

        # If the port was forced by the caller, punt
        die "$@\n" if $forced_port && $@;

        $server->port( random_port() );
    }
    while ( $@ );
    
    return 1; # This should be unreachable
}

### PRIVATE PACKAGE SUBROUTINE ###
#
# Generate a random port for the server to listen on
#

sub random_port { 30000 + int rand 10000 };

# Ensure that the server is stopped cleanly at exit
END { stop_server }

1;


lib/RPC/ExtDirect/Server/Patch/HTTPServerSimple.pm
package RPC::ExtDirect::Server::Patch::HTTPServerSimple;

use strict;
use warnings;
no  warnings 'redefine';

# This monkey patching is required for HTTP::Server::Simple <= 0.44;
# CGI.pm < 3.36 does not support HTTP_COOKIE environment variable with
# multiple values separated by commas instead of semicolons.
#
# The code is copied from HTTP::Server::Simple::CGI::Environment,
# with a fix applied.

use HTTP::Server::Simple::CGI::Environment;

sub HTTP::Server::Simple::CGI::Environment::header {
    my $self  = shift;
    my $tag   = shift;
    my $value = shift;

    $tag = uc($tag);
    $tag =~ s/^COOKIES$/COOKIE/;
    $tag =~ s/-/_/g;
    $tag = "HTTP_" . $tag
        unless $tag =~ m/^CONTENT_(?:LENGTH|TYPE)$/;

    if ( exists $ENV{$tag} ) {
        # This line is fixed
        $ENV{$tag} .= $tag eq 'HTTP_COOKIE' ? "; $value" : ", $value";
    }
    else {
        $ENV{$tag} = $value;
    }
}

1;



Makefile.PL
use 5.006000;
use ExtUtils::MakeMaker;

# Add the `devtest` target to run regression and POD tests in one go
sub MY::postamble {
    return <<'END';
devtest :
	REGRESSION_TESTS=1 POD_TESTS=1 $(MAKE) test

END
}

# Override `disttest` so it would behave as `devtest`
sub MY::dist_test {
    return <<'END';
disttest : distdir
	cd $(DISTVNAME) && $(ABSPERLRUN) Makefile.PL
	cd $(DISTVNAME) && $(MAKE) $(PASTHRU)
	cd $(DISTVNAME) && $(MAKE) devtest $(PASTHRU)

END
}

my $MM_VERSION  = $ExtUtils::MakeMaker::VERSION;
my $github_repo = 'https://github.com/nohuhu/RPC-ExtDirect-Server';

my %MAIN_REQ = (
    'RPC::ExtDirect'       => '3.21',
    'CGI::ExtDirect'       => '3.20',
    'HTTP::Server::Simple' => '0.44',
);

my %TEST_REQ = (
    'HTTP::Tiny' => '0.014',
    'Test::More' => '0.82',  # for explain()
);

WriteMakefile(
    NAME         => 'RPC::ExtDirect::Server',
    VERSION_FROM => 'lib/RPC/ExtDirect/Server.pm',
    ABSTRACT     => 'A tiny but capable pure Perl Ext.Direct server',
    AUTHOR       => 'Alex Tokarev <tokarev@cpan.org>',
    LICENSE      => 'perl',

    ($MM_VERSION >= 6.64
        ? (
            TEST_REQUIRES => \%TEST_REQ,
            PREREQ_PM     => \%MAIN_REQ,
        )
        : (
            ($MM_VERSION >= 6.5503
                ? (
                    BUILD_REQUIRES => \%TEST_REQ,
                    PREREQ_PM      => \%MAIN_REQ,
                )
                : (
                    PREREQ_PM => {
                        %TEST_REQ,
                        %MAIN_REQ,
                    },
                ),
            ),
        ),
    ),


    ($MM_VERSION >= 6.48 ? ( MIN_PERL_VERSION => 5.006000 ) : ()),

    ($MM_VERSION >= 6.46
        ? ( META_MERGE => {
                resources => {
                    bugtracker => "$github_repo/issues/",
                    repository => $github_repo,
                },
            },
        )
        : ()
    ),
);

